zproject/backends.py

```diff
@@ -920,44 +920,47 @@ class ExternalAuthResult:
     LOGIN_TOKEN_LENGTH = UserProfile.API_KEY_LENGTH
 
     def __init__(self, *, user_profile: Optional[UserProfile]=None,
-                 data_dict: ExternalAuthDataDict={},
+                 data_dict: Optional[ExternalAuthDataDict]=None,
                  login_token: Optional[str]=None,
                  delete_stored_data: bool=True) -> None:
+        if data_dict is None:
+            data_dict = {}
+
         if login_token is not None:
+            assert (not data_dict) and (user_profile is None), ("Passing in data_dict or user_profile " +
+                                                                "with login_token is disallowed.")
             self.instantiate_with_token(login_token, delete_stored_data)
-            return
         else:
-            self.data_dict = data_dict
-        self.user_profile = user_profile
-        if self.user_profile:
-            self._load_user_data_from_user_profile()
+            self.data_dict = cast(ExternalAuthDataDict, {**data_dict})
+            self.user_profile = user_profile
+
+        if self.user_profile is not None:
```

Comment:
It seems like we could have a method `auth_data_dict_for_user` that did this logic, and then we wouldn't need to have optional parameters for `data_dict` and `user_profile` in this function, and we could instead just always require `data_dict`, to keep things simple.  And then the caller can just call `auth_data_dict_for_user` as needed?

Reply:
@showell I agree that would be much nicer, but as things are `login_or_register_remote_user` does different stuff if it gets a `user_profile` as opposed to `user_profile=None` and some user data in the other arguments. Messing with that withing this refactoring commit would be undreadable, but I think this looks like a good direction for the follow-up after this commit though!